File: puzzles/medium/ruby/the-gift.rb
@creatures = gets.to_i # number of participants
@price = gets.to_i
budgets = Array.new()
total_budget = 0

@creatures.times do
  budget = gets.to_i
  budgets.push(budget)
  total_budget += budget
end

if total_budget < @price
  puts "IMPOSSIBLE"
else
  budgets = budgets.sort
  for i in 0..(@creatures - 1)
    contribution = (budgets[i] >= @price/(@creatures - i)) ? @price/(@creatures - i) : budgets[i]
    puts "#{contribution}"
    @price -= contribution
  end
end

File: puzzles/ts/the-gift/the-gift.ts
const nbParticipants: number = Number(readline());
let giftPrice = Number(readline());
let budgets: number[] = [];
let totalBudget = 0;

for (let i: number = 0; i < nbParticipants; i++) {
  const budget = Number(readline());
  budgets.push(budget);
  totalBudget += budget;
}

if (totalBudget < giftPrice) {
  console.log("IMPOSSIBLE");
} else {
  budgets.sort((b1, b2) => (b1 - b2));
  for (let i: number = 0; i < nbParticipants; i++) {
    let contribution = budgets[i];
    const maxContribution = Math.floor(giftPrice / (nbParticipants - i));
    if (contribution > maxContribution) {
      contribution = maxContribution;
    }
    console.log(contribution);
    giftPrice -= contribution;
  }
}
